Slurm worker manager

candicegjing opened the pull request:
Fixed #1019 .
I want to get some feedback for this PR in case I missed any details or use cases here.
### Overview
To submit a `slurm-batch` job, the user needs to run the following command to start the worker manager:
```
cl-worker-manager --server=http://localhost slurm-batch --nodelist=jagupard10
```
The Slurm batch worker manager will just run once without keep watching for future staged bundles. Hence, if the user wants to spawn a new Slurm worker, he/she has to run the same command another time. 

A successfully spawned worker job from `sc` will look like as follows:
```
stanford@Stanfords-MacBook-Pro codalab-worksheets % cl-worker-manager --server=http://localhost slurm-batch --nodelist=jagupard10
Namespace(cpus=1, dry_run=False, gpus=1, job_definition_name='codalab-slurm-worker', max_workers=10, memory_mb=2048, min_seconds_between_workers=60, min_workers=1, nodelist='jagupard10', once=True, partition='jag-standard', search=[], server='http://localhost', sleep_time=5, verbose=False, worker_idle_seconds=600, worker_manager_name='slurm-batch', worker_tag=None)
2020-05-20 17:45:39,693 Started worker manager.
2020-05-20 17:45:39,729 Staged bundles [state=staged]: 0xce9d53b2415e45c182a0693beed3847d 0xf311bf0cd69040cdb0e47cbb593ac3b7 0x4eb02dcc325244ff9b04c1931072bb2f 0x502fad7aabd343e99b89c71bf7485a71 0x0f91f28629bc4efcaeb753664a4aff20 0x2b7ffc86d4464c689b9f8d03179cb8c4 0x68a230ae604e4db2bb24cb1afbc3afa1 0x37f2595625794b6daace08b34930bff7 0x30c8ead85fda4358bac21973d8820e8a 0x090c800cdb594200818ffad8793aeef9
Get slurm jobs
2020-05-20 17:45:39,729 10 staged bundles (0 removed since last time), 0 worker jobs (min=1, max=10) (0 active, 0 pending)
2020-05-20 17:45:39,729 Want to launch a worker because we have 10 > 0 staged bundles
2020-05-20 17:45:39,729 Launch a worker because we are under the minimum (0 < 1)
2020-05-20 17:45:39,729 Starting a worker!
Slurm Batch Job Definition
#!/bin/bash

#SBATCH --cpus-per-task=3
#SBATCH --gres=gpu:1
#SBATCH --job-name=stanford-codalab-slurm-worker-3086758
#SBATCH --mem-per-cpu=2048
#SBATCH --nodelist=jagupard10
#SBATCH --ntasks-per-node=1
#SBATCH --open-mode=append
#SBATCH --output=stanford-codalab-slurm-worker-3086758.out
#SBATCH --partition=jag-standard
#SBATCH --time=10-0

srun --unbuffered cl-worker --server http://localhost --verbose --exit-when-idle --idle-seconds 600 --work-dir /Users/stanford/slurm-batch-scratch/4e524789af034496b1717e943ceed111 --id 4e524789af034496b1717e943ceed111 --network-prefix cl_worker_4e524789af034496b1717e943ceed111_network --pass-down-termination
2020-05-20 17:45:39,730 Saved the Slurm Batch Job Definition to /Users/stanford/slurm-batch-scratch/4e524789af034496b1717e943ceed111/stanford-codalab-slurm-worker-3086758.slurm
2020-05-20 17:45:39,733 Submitted job: sbatch stanford-codalab-slurm-worker-3086758.slurm
```

In the same time, I am adding more flexibility to Slurm job parameter.
Please feel free to suggest any use cases that I am missing here.

Commit:
Slurm worker manager

Commit:
Update logic

Commit:
Update logic

Commit:
Update logic

Commit:
Update format

Commit:
Update comments

Commit:
Update get_worker_jobs

Commit:
Update logic

Commit:
Add new line

Commit:
Update name

Commit:
Remove unnecessary lines

Commit:
Update variable naming

Commit:
Update logic

Commit:
Update logic

Commit:
Update functionality

Commit:
Update comments

Commit:
Clean up library

Commit:
Revert bad commit

Commit:
Update comments

Commit:
Update logic

Commit:
Update main

Commit:
Update logi

Commit:
Fix syntax

Commit:
Update logic

Commit:
Add return

Commit:
Update logic

Commit:
Update logic

Commit:
Update logic

Commit:
Update logic

Commit:
Add more logic

Commit:
Update details

Commit:
Update logic

Commit:
Add comments

Commit:
Revert changes

candicegjing committed:
Merge branch 'master' into slurm-worker-manager

candicegjing committed:
Merge branch 'master' into slurm-worker-manager

Commit:
Update logic

candicegjing commented:
@percyliang @nelson-liu This PR is tested on dev environment with the following cases:
1. Properly spawn the desired number of Slurm workers.
1. Jobs are able to run and finish once the allocated worker is ready.
1. Preemption on worker resources is working properly, including 
    1. jobs are able to send back to the `staged` state with `pass-down-termination` option specified
    1. local docker containers are stopped and removed as necessary  
1. Cancel a Slurm allocated worker resource is working properly, including 
    1. jobs can be stopped 
    1. local docker containers are stopped and removed as necessary


TODO
1. Testing of 1 job per worker will be done once the implementation of https://github.com/codalab/codalab-worksheets/issues/2287 is merged. 
1. Please feel free to suggest what else we need to add here.

Commit:
Add password-file as required

Commit:
Update code based on comments

Commit:
Update mem option

Commit:
Update name

nelson-liu commented:
Hm, so I'm trying to run the `slurmworkermanager` and test it under some real-world use cases. There isn't a very satisfying way to set the `work-dir`, though.

Currently, in `nlp-cl-worker`, I have:

```
CMD="${CMD} --work-dir ${HOSTNAME}/$(get-scr-with-most-space)/$USER-cl-worker"
```

i.e., the work-dir is only set once we get into the host. There isn't a very good way of doing this at the moment. The best thing I can think of (disclaimer: entirely a hack) is to have a custom script `cl-worker` that comes before the codalab-installed `cl-worker`, where the custom script defines the `--work-dir` based off of the current host and otherwise is just a pass-through to the codalab-installed `cl-worker`. Ideally, we'd like to be able to populate these fields with arbitrary values from the host, but I can't think of a clean way to do this...thoughts?

candicegjing commented:
> disclaimer

1. For populating `$(get-scr-with-most-space)`, what are the potential directories that we are allowed to use?
1. Since this is going to be Slurm specific, we can add this functionality to this class.
1. Another note is that we need to be able to keep track of where the directory is. Otherwise, with more and more directories being created, we might completely lose control of where bundles are stored.

nelson-liu commented:
I think I figured out a workaround (for now). For each worker node, I made symbolic links from /scr/scrX -> /$(hostname -a)/scrX, and a separate link /scr/scr-with-most-space -> /$(hostname -a)/(scrX that has the most empty space) . So, for now, I'll just set the `work-dir` to the static path `/scr/scr-with-most-space` . let's see if it holds up, and then we can optimize if it doesn't.

> For populating $(get-scr-with-most-space), what are the potential directories that we are allowed to use?

`/$(hostname -a)/scr*`

> Since this is going to be Slurm specific, we can add this functionality to this class.

Hmm..it's technically SAIL-cluster specific. Not sure how I feel about adding something so specific to this class in particular.

> Another note is that we need to be able to keep track of where the directory is. Otherwise, with more and more directories being created, we might completely lose control of where bundles are stored.

I'm not sure what you mean, could you elaborate?

Commit:
Add permission check

Commit:
Update dir

Commit:
Add worker dir

Commit:
Update password file

Commit:
Add codalab login logic

candicegjing committed:
Merge branch 'master' into slurm-worker-manager

Commit:
Update comments

Commit:
Merge branch 'slurm-worker-manager' of https://github.com/codalab/codalab-worksheets into slurm-worker-manager

Commit:
Update based on comments and refactoring

Commit:
Set partition as requirement CLI argument

Commit:
Remove period

candicegjing commented:
@nelson-liu I updated the code. In addition to what you requested, I refactored a bit more just to make the functionality more robust and easy to read. Please take a look and let me know what else we need to modify here.

Commit:
Update code

nelson-liu reviewed (approved):
LGTM, thanks for working on this!!

Commit:
Update comments

Commit:
Update default work directory
